tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,272 @@
+public:
+    int uniquePaths(int m, int n) {
+      // Make a 2-dimensional DP array with state i, j representing the number of ways to reach the cell (i, j)
+      // The problem can be solved using 1-dimensional DP array using space optimisation 
+      // (not implemented in favour of readablity)
+      vector<vector<int>> dp(m, vector<int> (n, 0));
+      // There is only one way to reach the topmost and leftmost square (base case)
+      dp[0][0] = 1;
+      for(int i = 0; i < m; i++){
+          for(int j = 0; j < n; j++){
+              // Transition: dp[i][j] = dp[i - 1][j] + dp[i][j - 1] 
+              if(i > 0) dp[i][j] += dp[i - 1][j];
+              if(j > 0) dp[i][j] += dp[i][j - 1];
+          }
+      }
+      return dp[m - 1][n - 1];
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+You can check the complete solution for this problem [here](../../solutions/0000-0099/unique-paths-medium)
+
+### Example #4 [2400 - Number of Ways to Reach a Position After Exactly k Steps](https://leetcode.com/problems/number-of-ways-to-reach-a-position-after-exactly-k-steps/)
+
+If you found this problem difficult, you can try [a similar problem](https://cses.fi/problemset/task/1746) in CSES Problem set.
+
+This is another problem where using DP can help as we are provided with a simple recurrence relation and are asked to find number of ways to arrive at some final state.
+
+We define $dp_{i, j}$ to define number of ways to reach $j^{th}$ position with $i$ moves, where $i$ iterates from $0$ to $k$ and $j$ represents the current position. 
+
+Then the recurrence relation can be written as $dp_{i, j}$ $=$ $dp_{i - 1, j - 1}$ $+$ $dp_{i - 1, j + 1}$
+
+Thus the time complexity of the solution is $O(k^2)$ with $O(k)$ space complexity.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int numberOfWays(int startPos, int endPos, int k) {
+        int mod = 1e9 + 7;
+        // Find the min and max position we can reach in k steps.
+        int min_pos = min(min(startPos, endPos) - (k - abs(startPos - endPos) + 1)/2, min(startPos, endPos));
+        int max_pos = max(max(startPos, endPos) + (k - abs(startPos - endPos) + 1)/2, max(startPos, endPos));
+        // Making DP table with to store from min position to max position 
+        vector<int> dp(max_pos - min_pos + 1, 0);
+        // Initialize the current location at startPos
+        dp[startPos - min_pos] = 1;
+        for(int i = 0; i < k; i++){
+            // Create a new DP array to store the values for next step
+            vector<int> new_dp(max_pos - min_pos + 1, 0);
+            for(int j = 0; j < (max_pos - min_pos + 1); j++){
+                if(j > 0) new_dp[j] += dp[j - 1];
+                if(j < (max_pos - min_pos)) new_dp[j] += dp[j + 1];
+                new_dp[j] %= mod;
+            }
+            // Update the DP array with the calculated values
+            dp = new_dp;
+        }
+        return dp[endPos - min_pos];
+    }
+};
+```
```

Comment:
same as above. here's my solution for your reference.

```cpp
struct comb {
 public:
  int MOD = 1e9 + 7;
  vector<long long> fac, finv, inv;
  comb(int mxN) {
    fac.resize(mxN);
    finv.resize(mxN);
    inv.resize(mxN);
    fac[0] = fac[1] = 1;
    finv[0] = finv[1] = 1;
    inv[1] = 1;
    for (int i = 1; i < mxN - 1; i++) {
      fac[i + 1] = fac[i] * (i + 1) % MOD;
      inv[i + 1] = MOD - inv[MOD % (i + 1)] * (MOD / (i + 1)) % MOD;
      finv[i + 1] = finv[i] * inv[i + 1] % MOD;
    }
  }

  long long ncr(long long n, long long k) {
    if (k < 0 || n < k) return 0;
    long long res = 1;
    res *= fac[n] * finv[k] % MOD * finv[n - k] % MOD;
    return res;
  }
};
class Solution {
public:
    int numberOfWays(int startPos, int endPos, int k) {
        int MOD = 1e9 + 7;
        // init comb
        // see constraints: 1 <= startPos, endPos, k <= 1000
        comb c = comb(1005);
        // if (startPos + endPos) and k have different parity, then return 0
        // otherwise we can use comb to get k choose (endPos - startPos + k) / 2
        // it can also be k choose k - (endPos - startPos + k) / 2)
        // assuming endPos >= startPos, and we know
        // left + right = k -- (1)
        // right - left = endPos - startPos -- (2)
        // (1) + (2)
        // left + right - left + right = k + endPos - startPos
        // 2 * right = endPos - startPos + k
        // right = (endPos - startPos + k) / 2
        // in other word, left would be k - (endPos - startPos + k) / 2)
        return (((startPos + endPos) ^ k) & 1) ? 0 : c.ncr(k, (endPos - startPos + k) / 2) % MOD;
    }
};
```